subprojects/frameworklauncher/yarn/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -201,7 +207,19 @@ public synchronized void stop(StopStatus stopStatus) {
     AggregateException ae = new AggregateException();
 
     // Stop AM's SubServices
+
     // No need to stop nmClient, since it may be time consuming to stop all Containers, leave it for RM.
+    try {
+      if (stopStatus.getNeedUnregister()) {
+        if (nmClient != null) {
+          LOGGER.logInfo("Trying to perform graceful shutdown and stop all the containers");
+          nmClient.stop();
```

Comment:
As comment said, it may be time consuming. We at least need a flag to skip the graceful shutdown.